src/backend/c-lisp.py

```diff
@@ -314,182 +323,289 @@ def get_literal_type(self, expr):
+        return expr[0] == "ptradd"
+
+    def compile(self, expr):
+        if len(expr) != 3:
+            raise CodegenError(f"Bad ptradd expression: {expr}")
+
+        offset = super().compile(expr[2])
+        ptr_name = self.ctx.scoped_lookup(expr[1])
+        ptr_type = self.ctx.symbol_types[ptr_name]
+        res_sym = random_label(CLISP_PREFIX)
+
+        instrs = offset.instructions + [
+            ["set", [res_sym, ptr_type], ["ptradd", ptr_name, offset.symbol]]
+        ]
+
+        return ExpressionResult(instrs, res_sym, ptr_type)
+
+
+class LoadExpression(Expression):
+    @classmethod
+    def is_valid_expr(cls, expr):
+        return expr[0] == "load"
+
+    def compile(self, expr):
+        if len(expr) != 2:
+            raise CodegenError(f"Bad load expression: {expr}")
+
+        ptr = super().compile(expr[1])
+        res_sym = random_label(CLISP_PREFIX)
+        instrs = ptr.instructions + [
+            ["set", [res_sym, ptr.typ[1]], ["load", ptr.symbol]]
+        ]
+        return ExpressionResult(instrs, res_sym, ptr.typ[1])
+
+
+class StoreExpression(Expression):
+    @classmethod
+    def is_valid_expr(cls, expr):
+        return expr[0] == "store"
+
+    def compile(self, expr):
+        if len(expr) != 3:
+            raise CodegenError(f"Bad store expression: {expr}")
+
+        ptr = super().compile(expr[1])
+        val = super().compile(expr[2])
+        if ptr.typ[1] != val.typ:
+            # TODO: Error test
+            raise CodegenError(f"Cannot store {val.typ} value to {ptr.typ}")
+
+        instrs = [
+            *ptr.instructions,
+            *val.instructions,
+            ["store", ptr.symbol, val.symbol],
+        ]
+
+        return ExpressionResult(instrs, val.symbol, val.typ)
+
+
+class AllocExpression(Expression):
+    @classmethod
+    def is_valid_expr(cls, expr):
+        return expr[0] == "alloc"
+
+    def compile(self, expr):
+        if len(expr) != 3:
+            raise CodegenError(f"Bad alloc expression: {expr}")
+
+        ptr_type = ["ptr", expr[1]]
+        res_sym = random_label(CLISP_PREFIX)
+        size = super().compile(expr[2])
+        instrs = [
+            *size.instructions,
+            ["set", [res_sym, ptr_type], ["alloc", size.symbol]],
+        ]
+        return ExpressionResult(instrs, res_sym, ptr_type)
+
+
+class CastExpression(Expression):
+    cast_ops = {
+        ## opcode -> operand type indicator, result type indicator
+        # Type conversion
+        "sitofp",
+        "fptosi",
+        "uitofp",
+        "fptoui",
+        "inttoptr",
+        "ptrtoint",
+        "sext",
+        "zext",
+        "trunc",
+        "fpext",
+        "fptrunc",
+        "bitcast",
+    }
+
+    @classmethod
+    def is_valid_expr(cls, expr):
+        return expr[0] in cls.cast_ops
+
+    def compile(self, expr):
+        opcode = expr[0]
+        if len(expr) != 3:
+            raise CodegenError(f"`{opcode}` takes exactly 2 operands")
+
+        # TODO: Type checking
```

Comment:
Add tests and type checking. I want to merge this.

Reply:
On it

Reply:
Almost done. Only inttoptr and ptrtoint remain

Reply:
Tests for all cast operators and type-checking for casting is done.